File: backend/test/multiplayer.e2e.test.ts
import http from 'node:http';
import type { AddressInfo } from 'node:net';

import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest';
import request from 'supertest';
import { io as createClient, type Socket as ClientSocket } from 'socket.io-client';

import { attachSocketServer, stopSocketServer } from '../src/socket.js';
import { roomKeyFor } from '../src/multiplayer/cache.js';
import { redis } from '../src/db/redis.js';
import { closeInfra, getApp, prepareDatabase, resetAuthState } from './helpers.js';

const TOTAL_ROUNDS = 5;

let server: http.Server;
let serverUrl: string;
const liveClients: ClientSocket[] = [];

// 服务端中止对局（mp:error，如题目池为空）时立即拒绝挂起的等待，
// 让失败信息直接指向根因，而不是空等 8 秒后报"事件超时"
function attachAbortHandler(client: ClientSocket, reject: (reason: Error) => void): void {
    client.once('mp:error', (data: unknown) => {
        const message =
            typeof data === 'object' && data !== null && 'message' in data
                ? String((data as { message: unknown }).message)
                : '未知错误';
        reject(new Error(`服务端中止对局：${message}`));
    });
}

function waitFor(client: ClientSocket, event: string, timeoutMs = 8000): Promise<unknown> {
    return new Promise((resolve, reject) => {
        attachAbortHandler(client, reject);
        const timer = setTimeout(() => {
            client.off(event, handler);
            reject(new Error(`等待事件 ${event} 超时`));
        }, timeoutMs);
        const handler = (data: unknown): void => {
            clearTimeout(timer);
            resolve(data);
        };
        client.on(event, handler);
    });
}

// 事件缓冲：注册即收，消费时再取。服务端可能在客户端挂上监听器之前就下发事件
//（例如 mp:matched 之后紧跟的 mp:round），直接 waitFor 会因监听注册太晚而永久丢失
function makeEventBuffer<T>(client: ClientSocket, event: string): { next(timeoutMs?: number): Promise<T> } {
    const queue: T[] = [];
    const waiters: Array<{ resolve: (value: T) => void; timer: NodeJS.Timeout }> = [];
    client.on(event, (data: T) => {
        const waiter = waiters.shift();
        if (waiter !== undefined) {
            clearTimeout(waiter.timer);
            waiter.resolve(data);
        } else {
            queue.push(data);
        }
    });
    return {
        next(timeoutMs = 8000): Promise<T> {
            const queued = queue.shift();
            if (queued !== undefined) {
                return Promise.resolve(queued);
            }
            return new Promise((resolve, reject) => {
                attachAbortHandler(client, reject);
                const timer = setTimeout(() => reject(new Error(`等待事件 ${event} 超时`)), timeoutMs);
                waiters.push({ resolve, timer });
            });
        },
    };
}

function waitForConnect(client: ClientSocket): Promise<void> {
    return new Promise((resolve, reject) => {
        client.on('connect', () => resolve());
        client.on('connect_error', (err: Error) => reject(new Error(`连接失败：${err.message}`)));
    });
}

function connectClient(token: string): ClientSocket {
    const client = createClient(serverUrl, {
        auth: { token },
        transports: ['websocket'],
        forceNew: true,
        reconnection: false,
    });
    liveClients.push(client);
    return client;
}

async function createGuestToken(): Promise<string> {
    const response = await request(getApp()).post('/api/auth/guest');
    expect(response.status).toBe(201);
    return response.body.guestToken as string;
}

// 服务端不再向客户端下发本回合答案坐标（mp:round 仅含渲染街景所需字段），
// 测试需像权威计分那样从房间缓存取得真实坐标来构造"精确命中"与"远离"的猜测
async function currentRoundLocation(roomId: string): Promise<{ lat: number; lng: number } | null> {
    const raw = await redis.hget(roomKeyFor(roomId), 'location');
    if (raw === undefined || raw === null || raw === '') {
        return null;
    }
    return JSON.parse(raw) as { lat: number; lng: number };
}

// 与对局协议联动的完整 5 回合流程：等待 round -> 提交答案 -> 等待 roundEnd
// near = true 的选手提交与答案完全相同的坐标（0km -> 满分），far 的选手提交对跖点（几乎 0 分）
async function playFullMatch(client: ClientSocket, near: boolean): Promise<{ rankings: unknown }> {
    // 在等待 matched 之前就注册 round/roundEnd 缓冲，杜绝服务端先于监听器下发事件导致的丢事件
    const rounds = makeEventBuffer<{ roundIndex: number; location: Record<string, unknown> }>(client, 'mp:round');
    const roundEnds = makeEventBuffer<unknown>(client, 'mp:roundEnd');
    const matched = (await waitFor(client, 'mp:matched')) as { roomId: string };
    for (let roundIndex = 0; roundIndex < TOTAL_ROUNDS; roundIndex++) {
        const roundData = await rounds.next();
        expect(roundData.roundIndex).toBe(roundIndex);
        expect(roundData.location.name).toBeUndefined();
        // 安全回归断言：题目坐标绝不随 mp:round 下发
        expect(roundData.location.lat).toBeUndefined();
        expect(roundData.location.lng).toBeUndefined();
        const location = await currentRoundLocation(matched.roomId);
        if (location === null) {
            throw new Error('无法读取本回合答案坐标');
        }
        client.emit(
            'mp:answer',
            near
                ? { guessLat: location.lat, guessLng: location.lng, roundIndex }
                : { guessLat: -location.lat, guessLng: (location.lng >= 0 ? location.lng - 180 : location.lng + 180), roundIndex }
        );
        await roundEnds.next();
    }
    return (await waitFor(client, 'mp:finished')) as { rankings: unknown };
}

beforeAll(async () => {
    await prepareDatabase();
    server = http.createServer(getApp());
    attachSocketServer(server);
    await new Promise<void>((resolve) => server.listen(0, resolve));
    const address = server.address() as AddressInfo;
    serverUrl = `http://localhost:${address.port}`;
});

beforeEach(async () => {
    await resetAuthState();
});

afterEach(() => {
    for (const client of liveClients) {
        client.disconnect();
    }
    liveClients.length = 0;
});

afterAll(async () => {
    stopSocketServer();
    await new Promise<void>((resolve) => server.close(() => resolve()));
    await closeInfra();
});

describe('Socket.IO 多人对战', () => {
    it('双人完成 5 回合对局并按服务端计分排出名次', async () => {
        const tokenA = await createGuestToken();
        const tokenB = await createGuestToken();
        const clientA = connectClient(tokenA);
        const clientB = connectClient(tokenB);
        await Promise.all([waitForConnect(clientA), waitForConnect(clientB)]);

        clientA.emit('mp:join', { mode: 'classic' });
        clientB.emit('mp:join', { mode: 'classic' });

        const [finishedA, finishedB] = await Promise.all([
            playFullMatch(clientA, true),
            playFullMatch(clientB, false),
        ]);

        const rankingsA = finishedA.rankings as Array<{ playerId: string; username: string; totalScore: number }>;
        const rankingsB = finishedB.rankings as Array<{ playerId: string; username: string; totalScore: number }>;
        expect(rankingsA).toHaveLength(2);
        expect(rankingsA[0].totalScore).toBeGreaterThan(rankingsA[1].totalScore);
        expect(rankingsA[0]).toEqual(rankingsB[0]);
        expect(rankingsA[0].totalScore).toBe(TOTAL_ROUNDS * 5000);
        expect(rankingsA[1].totalScore).toBe(0);

        clientA.disconnect();
        clientB.disconnect();
    });

    it('对局结束后成绩写入 game_results（duel 模式）', async () => {
        const tokenA = await createGuestToken();
        const tokenB = await createGuestToken();
        const clientA = connectClient(tokenA);
        const clientB = connectClient(tokenB);
        await Promise.all([waitForConnect(clientA), waitForConnect(clientB)]);

        clientA.emit('mp:join', { mode: 'classic' });
        clientB.emit('mp:join', { mode: 'classic' });
        await Promise.all([playFullMatch(clientA, true), playFullMatch(clientB, false)]);
        clientA.disconnect();
        clientB.disconnect();

        const recentA = await request(getApp()).get('/api/games/recent').set('Authorization', `Bearer ${tokenA}`);
        expect(recentA.status).toBe(200);
        expect(recentA.body.games).toHaveLength(1);
        expect(recentA.body.games[0].mode).toBe('duel');
        expect(recentA.body.games[0].rounds).toHaveLength(TOTAL_ROUNDS);
        expect(recentA.body.games[0].totalScore).toBe(TOTAL_ROUNDS * 5000);
    });

    it('未携带有效令牌的连接被拒绝', async () => {
        const client = connectClient('invalid-token');
        await expect(waitForConnect(client)).rejects.toThrow('连接失败');
        client.disconnect();
    });
});

File: backend/test/multiplayer_e2e_test.go
package e2e

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"testing"
	"time"
)

// eioClient is a minimal Engine.IO v4 polling client used to exercise the
// server exactly like the real socket.io 4.8.1 client.
type eioClient struct {
	t       *testing.T
	baseURL string
	sid     string
	recv    chan string
	httpc   *http.Client
	done    chan struct{}
	ctx     context.Context
	cancel  context.CancelFunc
}

func newEIOClient(t *testing.T, baseURL string) *eioClient {
	t.Helper()
	ctx, cancel := context.WithCancel(context.Background())
	client := &eioClient{
		t:       t,
		baseURL: baseURL,
		recv:    make(chan string, 64),
		httpc:   &http.Client{},
		done:    make(chan struct{}),
		ctx:     ctx,
		cancel:  cancel,
	}
	// Handshake.
	body := client.get("/socket.io/?EIO=4&transport=polling&t=1")
	if len(body) < 2 || body[0] != '0' {
		t.Fatalf("unexpected handshake: %q", body)
	}
	var handshake struct {
		Sid string `json:"sid"`
	}
	if err := json.Unmarshal([]byte(body[1:]), &handshake); err != nil {
		t.Fatalf("parse handshake: %v (%q)", err, body)
	}
	client.sid = handshake.Sid
	t.Cleanup(client.close)
	go client.pollLoop()
	return client
}

// get performs one polling GET. On transport error or non-200 it returns ""
// (the caller retries); t.Fatalf is only called while the client is still
// active so background polls never panic after the test has completed.
func (c *eioClient) get(url string) string {
	c.t.Helper()
	req, err := http.NewRequestWithContext(c.ctx, http.MethodGet, c.baseURL+url, nil)
	if err != nil {
		return ""
	}
	resp, err := c.httpc.Do(req)
	if err != nil {
		select {
		case <-c.done:
		default:
			c.t.Fatalf("poll failed: %v", err)
		}
		return ""
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		_, _ = io.Copy(io.Discard, resp.Body)
		return ""
	}
	raw, _ := io.ReadAll(resp.Body)
	return string(raw)
}

func (c *eioClient) post(frame string) {
	c.t.Helper()
	req, err := http.NewRequestWithContext(c.ctx, http.MethodPost, c.baseURL+"/socket.io/?EIO=4&transport=polling&sid="+c.sid,
		strings.NewReader(frame))
	if err != nil {
		return
	}
	resp, err := c.httpc.Do(req)
	if err != nil {
		select {
		case <-c.done:
		default:
			c.t.Fatalf("post failed: %v", err)
		}
		return
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, resp.Body)
}

// pollLoop keeps a long-poll open; engine.io frames are pushed to recv and
// pings answered with pongs. It returns as soon as the client is closed so
// that t.Cleanup never leaves stray goroutines panicking.
func (c *eioClient) pollLoop() {
	for {
		select {
		case <-c.done:
			return
		default:
		}
		raw := c.get("/socket.io/?EIO=4&transport=polling&sid=" + c.sid)
		if raw == "" {
			select {
			case <-c.done:
				return
			default:
			}
			time.Sleep(20 * time.Millisecond)
			continue
		}
		for _, frame := range strings.Split(raw, "\x1e") {
			if frame == "" {
				continue
			}
			if frame[0] == '2' { // server ping → pong
				c.post("3")
				continue
			}
			select {
			case c.recv <- frame:
			default:
			}
		}
	}
}

// connect sends the socket.io CONNECT packet with auth and waits for the ack.
func (c *eioClient) connect(token string) {
	c.t.Helper()
	payload, _ := json.Marshal(map[string]string{"token": token})
	c.post("40" + string(payload))
	if !c.expectPrefix("40", 3*time.Second) {
		c.t.Fatal("socket.io connect ack not received")
	}
}

// emit sends a socket.io EVENT packet.
func (c *eioClient) emit(event string, payload any) {
	c.t.Helper()
	args := []any{event}
	if payload != nil {
		args = append(args, payload)
	}
	raw, _ := json.Marshal(args)
	c.post("42" + string(raw))
}

// expectEvent waits for an event frame `42["name",...]` matching the event
// name and returns its arguments as JSON bytes.
func (c *eioClient) expectEvent(name string, timeout time.Duration) []byte {
	c.t.Helper()
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		select {
		case frame := <-c.recv:
			if !strings.HasPrefix(frame, "42") {
				continue
			}
			var parts []json.RawMessage
			if err := json.Unmarshal([]byte(frame[2:]), &parts); err != nil || len(parts) < 1 {
				continue
			}
			var eventName string
			if err := json.Unmarshal(parts[0], &eventName); err != nil || eventName != name {
				continue
			}
			if len(parts) > 1 {
				return parts[1]
			}
			// Event received with no arguments: return a non-nil empty
			// payload so callers can distinguish "found" from a timeout.
			return []byte{}
		case <-c.done:
			return nil
		}
	}
	return nil
}

func (c *eioClient) expectPrefix(prefix string, timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		select {
		case frame := <-c.recv:
			if strings.HasPrefix(frame, prefix) {
				return true
			}
		case <-c.done:
			return false
		}
	}
	return false
}

func (c *eioClient) close() {
	c.cancel() // abort any in-flight poll/post
	select {
	case <-c.done:
	default:
		close(c.done)
	}
}

func TestMultiplayerDuel(t *testing.T) {
	e := newTestEnv(t)
	e.seedLocations(50)

	guestA := e.request(t, http.MethodPost, "/api/auth/guest", "", nil)
	guestB := e.request(t, http.MethodPost, "/api/auth/guest", "", nil)

	clientA := newEIOClient(t, e.ts.URL)
	clientB := newEIOClient(t, e.ts.URL)
	clientA.connect(guestA.str("guestToken"))
	clientB.connect(guestB.str("guestToken"))

	t.Run("join queues both players", func(t *testing.T) {
		clientA.emit("mp:join", map[string]any{"mode": "duel"})
		clientB.emit("mp:join", map[string]any{"mode": "duel"})

		if raw := clientA.expectEvent("mp:queued", 3*time.Second); raw == nil {
			t.Fatal("player A never queued")
		}
		if raw := clientB.expectEvent("mp:queued", 3*time.Second); raw == nil {
			t.Fatal("player B never queued")
		}
	})

	var roomID string
	var opponentB string
	t.Run("matchmaker pairs the players", func(t *testing.T) {
		rawA := clientA.expectEvent("mp:matched", 5*time.Second)
		if rawA == nil {
			t.Fatal("player A never matched")
		}
		rawB := clientB.expectEvent("mp:matched", 5*time.Second)
		if rawB == nil {
			t.Fatal("player B never matched")
		}
		var matchedA struct {
			RoomID           string `json:"roomId"`
			Mode             string `json:"mode"`
			OpponentUsername string `json:"opponentUsername"`
		}
		if err := json.Unmarshal(rawA, &matchedA); err != nil {
			t.Fatalf("parse matched: %v", err)
		}
		roomID = matchedA.RoomID
		opponentB = matchedA.OpponentUsername
		if matchedA.Mode != "duel" {
			t.Fatalf("unexpected mode %s", matchedA.Mode)
		}
	})

	t.Run("five rounds with answers settle and finish", func(t *testing.T) {
		myID := guestA.str("guestId")
		for round := 0; round < 5; round++ {
			roundA := clientA.expectEvent("mp:round", 5*time.Second)
			roundB := clientB.expectEvent("mp:round", 5*time.Second)
			if roundA == nil || roundB == nil {
				t.Fatalf("round %d never started", round)
			}
			var roundPayload struct {
				RoundIndex  int `json:"roundIndex"`
				TotalRounds int `json:"totalRounds"`
				TimeLimitMs int `json:"timeLimitMs"`
				Location    struct {
					PanoramaURL string `json:"panoramaUrl"`
					MapillaryID string `json:"mapillaryId"`
				} `json:"location"`
			}
			if err := json.Unmarshal(roundA, &roundPayload); err != nil {
				t.Fatalf("parse round: %v", err)
			}
			if roundPayload.RoundIndex != round || roundPayload.TotalRounds != 5 ||
				roundPayload.TimeLimitMs != 60000 {
				t.Fatalf("unexpected round payload %+v", roundPayload)
			}
			if roundPayload.Location.MapillaryID != "" || roundPayload.Location.PanoramaURL != "" {
				// The round must NOT leak the answer coordinates.
				var leak struct {
					Location map[string]any `json:"location"`
				}
				_ = json.Unmarshal(roundA, &leak)
				if _, hasLat := leak.Location["lat"]; hasLat {
					t.Fatal("mp:round leaked answer coordinates")
				}
			}

			answer := map[string]any{
				"guessLat":   30 + float64(round),
				"guessLng":   100 + float64(round),
				"roundIndex": round,
			}
			clientA.emit("mp:answer", answer)
			clientB.emit("mp:answer", answer)

			endA := clientA.expectEvent("mp:roundEnd", 5*time.Second)
			endB := clientB.expectEvent("mp:roundEnd", 5*time.Second)
			if endA == nil || endB == nil {
				t.Fatalf("round %d never ended", round)
			}
			var endPayload struct {
				RoundIndex int `json:"roundIndex"`
				Answer     struct {
					Name string  `json:"name"`
					Lat  float64 `json:"lat"`
					Lng  float64 `json:"lng"`
				} `json:"answer"`
				Results []struct {
					PlayerID   string   `json:"playerId"`
					DistanceKm *float64 `json:"distanceKm"`
					Score      int      `json:"score"`
				} `json:"results"`
			}
			if err := json.Unmarshal(endA, &endPayload); err != nil {
				t.Fatalf("parse roundEnd: %v", err)
			}
			if endPayload.RoundIndex != round || len(endPayload.Results) != 2 {
				t.Fatalf("unexpected roundEnd %+v", endPayload)
			}
			if endPayload.Answer.Lat == 0 && endPayload.Answer.Lng == 0 {
				t.Fatal("roundEnd must reveal the answer")
			}
		}

		finished := clientA.expectEvent("mp:finished", 5*time.Second)
		if finished == nil {
			t.Fatal("player A never saw mp:finished")
		}
		if clientB.expectEvent("mp:finished", 5*time.Second) == nil {
			t.Fatal("player B never saw mp:finished")
		}
		var finishPayload struct {
			Rankings []struct {
				PlayerID   string `json:"playerId"`
				Username   string `json:"username"`
				TotalScore int    `json:"totalScore"`
			} `json:"rankings"`
		}
		if err := json.Unmarshal(finished, &finishPayload); err != nil {
			t.Fatalf("parse finished: %v", err)
		}
		if len(finishPayload.Rankings) != 2 {
			t.Fatalf("expected 2 rankings, got %d", len(finishPayload.Rankings))
		}
		// Both players answered identically, so scores must match.
		if finishPayload.Rankings[0].TotalScore != finishPayload.Rankings[1].TotalScore {
			t.Fatalf("identical answers should tie: %+v", finishPayload.Rankings)
		}
		_ = myID
		_ = roomID
		_ = opponentB
	})

	t.Run("duel games persisted", func(t *testing.T) {
		deadline := time.Now().Add(5 * time.Second)
		var count int
		for time.Now().Before(deadline) {
			_ = e.conn.QueryRow(`SELECT COUNT(*) FROM game_results WHERE mode = 'duel'`).Scan(&count)
			if count == 2 {
				break
			}
			time.Sleep(50 * time.Millisecond)
		}
		if count != 2 {
			t.Fatalf("expected 2 duel games persisted, got %d", count)
		}
	})

	t.Run("join before match leaves queue", func(t *testing.T) {
		guestC := e.request(t, http.MethodPost, "/api/auth/guest", "", nil)
		clientC := newEIOClient(t, e.ts.URL)
		clientC.connect(guestC.str("guestToken"))
		clientC.emit("mp:join", map[string]any{"mode": "duel"})
		if clientC.expectEvent("mp:queued", 3*time.Second) == nil {
			t.Fatal("player C never queued")
		}
		clientC.emit("mp:leave", nil)
		if clientC.expectEvent("mp:leftQueue", 3*time.Second) == nil {
			t.Fatal("player C never left the queue")
		}
	})
}

func TestMultiplayerAuthRejection(t *testing.T) {
	e := newTestEnv(t)

	t.Run("connect without token rejected", func(t *testing.T) {
		client := newEIOClient(t, e.ts.URL)
		payload, _ := json.Marshal(map[string]string{})
		client.post("40" + string(payload))
		if !client.expectPrefix("44", 3*time.Second) {
			t.Fatal("expected connect error 44")
		}
		client.close()
	})

	t.Run("connect with forged token rejected", func(t *testing.T) {
		client := newEIOClient(t, e.ts.URL)
		payload, _ := json.Marshal(map[string]string{"token": "forged"})
		client.post("40" + string(payload))
		if !client.expectPrefix("44", 3*time.Second) {
			t.Fatal("expected connect error 44")
		}
		client.close()
	})
}
